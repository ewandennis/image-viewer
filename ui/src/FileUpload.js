import React, { useCallback } from 'react';
import { DropzoneArea } from "material-ui-dropzone";
import { maxFileSizeBytes } from "./config";

export default function FileUpload({ onUploadFile }) {
  const handleChange = useCallback(
    fileList => {
      if (fileList && fileList.length > 0) {
        onUploadFile(fileList[0]);
      }
    },
    [onUploadFile]
  );

  return (
    <DropzoneArea
      acceptedFiles={["image/*"]}
      filesLimit={1}
      maxFileSize={maxFileSizeBytes}
      showAlerts={false}
      showPreviews={false}
      showPreviewsInDropzone={false}
      onChange={handleChange}
    />
  );
}
