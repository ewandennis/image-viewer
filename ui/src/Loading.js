import React from 'react';
import CircularProgress from "@material-ui/core/CircularProgress";

const Loading = () => (
  <div>
    <CircularProgress />
  </div>
);

export default Loading;
