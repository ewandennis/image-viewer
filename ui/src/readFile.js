const readFile = file => {
  return new Promise((resolve, reject) => {
    const reader = new FileReader();
    reader.onload = event => {
      resolve({ name: file.name, content: event.target.result });
    };
    reader.onerror = err => {
      reader.abort();
      reject(err);
    };
    reader.readAsBinaryString(file);
  });
};

export default readFile;
