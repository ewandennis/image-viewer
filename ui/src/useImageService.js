import React from 'react';
import socketIoClient from 'socket.io-client';
import { webSocketEndpoint } from './config';

// TODO move this out of top-level scope
// TODO close connection cleanly on app unmount?
const client = socketIoClient(webSocketEndpoint);

/*
 * Provide access to an shared image list.
 * 
 * API service maintains the source of truth.
 * This client uses a websocket to keep a local copy in sync.
 * 
 * @typedef {Object<string, any>} ImageService
 * @property {string} images The current image list
 * @property {function} uploadImage A way to upload new images
 * @param {string} endpoint API service websocket endpoint
 * @returns {ImageService}
 */
export default function useImageService(endpoint) {
  const [images, setImages] = React.useState(null);

  const uploadImage = (filename, image) => {
    client.emit('uploadImage', { filename, image });
  };

  React.useEffect(() => {
    const onImageList = imageList => {
      // TODO validate msg
      console.log(imageList);
      setImages(imageList);
    };

    const onImageAdded = image => {
      // TODO validate msg
      setImages([...images, image]);
    };

    client.on('imageList', onImageList);
    client.on('imageAdded', onImageAdded);

    if (images === null) {
      client.emit('listImages');
    }

    return () => {
      client.removeEventListener('imageList', onImageList);
      client.removeEventListener('imageAdded', onImageAdded);
    };
  }, [images]);

  return { images, uploadImage };
}