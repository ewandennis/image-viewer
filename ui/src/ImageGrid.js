import React, { useState, useCallback } from "react";

import Grid from "@material-ui/core/Grid";
import GridList from "@material-ui/core/GridList";
import GridListTile from "@material-ui/core/GridListTile";

import Loading from "./Loading";
import FileUpload from "./FileUpload";
import useImageService from "./useImageService";
import socketIoClient from 'socket.io-client';
import { webSocketEndpoint } from "./config";
import readFile from "./readFile";

const extractFileExtension = (filename) => filename.split('.').pop();

export default function ImageGrid() {
  const [ client, _ ] = useState(socketIoClient(webSocketEndpoint)); // eslint-disable-line no-unused-vars
  const { images, uploadImage } = useImageService(client);
  const [uploading, setUploading] = useState(false);
  const handleUpload = useCallback(
    file => {
      setUploading(true);
      return readFile(file)
        .then(({ name, content }) => {
          uploadImage({ filename: name , image: btoa(content) });
          setUploading(false);
        })
        .catch(err => {
          // TODO error UI
          console.error(err);
          setUploading(false);
        });
    },
    [setUploading, uploadImage]
  );

  if (images === null) {
    return <Loading />;
  }

  return (
    <Grid container justify="center" spacing={5}>
      <Grid item xs={8}>
        {uploading ? <Loading /> : <FileUpload onUploadFile={handleUpload} />}
      </Grid>
      <Grid item xs={12}>
        <GridList cols={4} spacing={10}>
          {images.map(img => (
            <GridListTile key={img.filename}>
              <img src={`data:image/${extractFileExtension(img.filename)};base64,${img.image}`} alt={img.filename} />
            </GridListTile>
          ))}
        </GridList>
      </Grid>
    </Grid>
  );
}
