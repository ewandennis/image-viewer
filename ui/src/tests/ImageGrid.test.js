import React from 'react';
import { shallow } from 'enzyme';
import path from 'path';
import ImageGrid from '../ImageGrid';
import Loading from '../Loading';
import FileUpload from "../FileUpload";
import useImageService from '../useImageService';
import readFile from '../readFile';

import { mockImageList } from './mockImageList';

jest.mock('socket.io-client');
jest.mock('../useImageService');
jest.mock('../readFile');


describe('ImageGrid', () => {
  const subject = () => shallow(<ImageGrid />);

  it('renders', () => {
    useImageService.mockReturnValue({ images: mockImageList, uploadImage: jest.fn() });
    expect(subject()).toMatchSnapshot();
  });

  it('renders while loading', () => {
    useImageService.mockReturnValue({ images: null, uploadImage: jest.fn() });
    expect(subject().find(Loading)).toExist();
  })

  it('renders an empty image list', () => {
    useImageService.mockReturnValueOnce({ images: [], uploadImage: jest.fn() });
    const wrapper = subject();
    expect(wrapper.find(Loading)).not.toExist();
    expect(wrapper.find('img')).not.toExist();
  });

  it('renders images', () => {
    useImageService.mockReturnValueOnce({ images: mockImageList, uploadImage: jest.fn() });
    const wrapper = subject();
    expect(wrapper.find(Loading)).not.toExist();
    expect(wrapper.find('img')).toHaveLength(mockImageList.length);
  });

  it('correctly uses image extension', () => {
    useImageService.mockReturnValueOnce({ images: mockImageList, uploadImage: jest.fn() });
    const wrapper = subject();
    expect(wrapper.find(Loading)).not.toExist();
    const images = wrapper.find('img');
    images.forEach((image, imageIdx) => {
      const expectedExtension = path.extname(mockImageList[imageIdx].filename).slice(1);
      const regExStr = `^data:image\/${expectedExtension}`;
      expect(image.prop('src')).toMatch(new RegExp(regExStr));
    });
  });

  it('accepts uploads', () => {
    const uploadImage = jest.fn();
    const file = { name: 'readmeplz.png' };
    const fileContents = { name: file.name, content: 'binary-image-bits-here' };
    useImageService.mockReturnValueOnce({ images: [], uploadImage });
    readFile.mockResolvedValue(fileContents);
    const wrapper = subject();
    return wrapper.find(FileUpload).prop('onUploadFile')(file)
    .then(() => {
      expect(uploadImage).toHaveBeenCalledTimes(1);
      expect(uploadImage).toHaveBeenCalledWith({ filename: file.name, image: btoa(fileContents.content) });
    });
  });
});
