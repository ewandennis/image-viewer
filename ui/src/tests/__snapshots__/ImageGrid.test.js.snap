// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ImageGrid renders 1`] = `
<WithStyles(ForwardRef(Grid))
  container={true}
  justify="center"
  spacing={5}
>
  <WithStyles(ForwardRef(Grid))
    item={true}
    xs={8}
  >
    <FileUpload
      onUploadFile={[Function]}
    />
  </WithStyles(ForwardRef(Grid))>
  <WithStyles(ForwardRef(Grid))
    item={true}
    xs={12}
  >
    <WithStyles(ForwardRef(GridList))
      cols={4}
      spacing={10}
    >
      <WithStyles(ForwardRef(GridListTile))
        key="pic.jpg"
      >
        <img
          alt="pic.jpg"
          src="data:image/jpg;base64,still-not-very-base64-encoded"
        />
      </WithStyles(ForwardRef(GridListTile))>
      <WithStyles(ForwardRef(GridListTile))
        key="photo.gif"
      >
        <img
          alt="photo.gif"
          src="data:image/gif;base64,still-not-very-base64-encoded"
        />
      </WithStyles(ForwardRef(GridListTile))>
    </WithStyles(ForwardRef(GridList))>
  </WithStyles(ForwardRef(Grid))>
</WithStyles(ForwardRef(Grid))>
`;
